### ISSUE TITLE
apfloat bug: incorrect handling of infinities in FMA

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#100233]
The Rust port of LLVM's apfloat has a bug in its FMA (fused-multiply-add) implementation. See https://github.com/rust-lang/miri/issues/2468#issuecomment-1207310464 for details, but basically it fails this assertion:
```rust
    let neg_inf: f32 = f32::NEG_INFINITY;
    assert_eq!((-3.2f32).mul_add(2.4, neg_inf), neg_inf);
```

AFAIK we don't use apfloat FMA anywhere (Miri stopped using them to work around this bug), but this should still be tracked, and will need to be fixed before we can allow FMA in `const`.

**Comment by eddyb** [comment]
Copying the relevant parts here (it's a clumsy mistake I made when porting):
> * LLVM's `APFloat` has [an unconditional sign multiplication first](https://github.com/llvm/llvm-project/blob/5bbe452e75d464d02547cef0ea2919c53b693f54/llvm/lib/Support/APFloat.cpp#L1985-L1986)
> * `rustc_apfloat` instead [only does it outside of handling infinities etc.](https://github.com/rust-lang/rust/blob/master/compiler/rustc_apfloat/src/ieee.rs#L850-L851)
>
> Seems like this bug was there from the start (of the port), I don't really understand how it happened (but obviously I screwed up the otherwise-simple early-return control-flow refactor) - what worries me a bit more is that I heavily relied on tests and [there's barely any for FMA](https://github.com/rust-lang/rust/blob/2befdefdda0f2bc5ff93f7f95633c98841409098/compiler/rustc_apfloat/tests/ieee.rs#L494-L567) (they're all finite AFAICT!).
And it's not like I lost some tests during porting, [they only had that many](https://github.com/llvm/llvm-project/blob/5bbe452e75d464d02547cef0ea2919c53b693f54/llvm/unittests/ADT/APFloatTest.cpp#L470-L561).

**Comment by eddyb** [comment]
> ... (it's a clumsy mistake I made when porting):
> * LLVM's `APFloat` has [an unconditional sign multiplication first](https://github.com/llvm/llvm-project/blob/5bbe452e75d464d02547cef0ea2919c53b693f54/llvm/lib/Support/APFloat.cpp#L1985-L1986)
> * `rustc_apfloat` instead [only does it outside of handling infinities etc.](https://github.com/rust-lang/rust/blob/master/compiler/rustc_apfloat/src/ieee.rs#L850-L851)

Turns out it's *not* that simple. I'm not actually doing the same thing as the C++ code, but instead I'm bypassing their `multiplySpecials` distinction entirely and using the normal multiplication facilities.

So it's more likely a case of missing some subtler special case because of the odd refactor.

<sub>(I'm not sure entirely why I did it that way, but FMA is at least special in one way: it needs more precision, and `[Limb; 2]` instead of `[Limb; 1]`, which `multiplySignificand` handles in the C++ version by whether `addend` is NULL, i.e. it has FMA and non-FMA modes, but I inlined `multiplySignificand` to remove the dynamic choice which was handled in the C++ code with *dynamic heap allocations*)</sub>
